lib/kang.js

```diff
@@ -80,20 +85,42 @@ function knMakeSource(str)
 		str = str.substr(0, c);
 	}
 
-	c = str.indexOf(':');
-	if (c == -1) {
-		host = str;
+	if ((matched = IPV4_HOST.exec(str)) !== null) {
+		host = matched[1];
+		portstr = matched[2];
+		if (!mod_net.isIPv4(host)) {
+			throw new Error('bad IPv4 address: ' +
+			    JSON.stringify(host));
+		}
+	} else if ((matched = IPV6_HOST.exec(str)) !== null) {
+		host = matched[1];
+		portstr = matched[2];
+		if (!mod_net.isIPv6(host)) {
+			throw new Error('bad IPv6 address: ' +
+			    JSON.stringify(host));
+		}
+	} else if ((matched = DOMAIN_HOST.exec(str)) !== null) {
+		host = matched[1];
+		portstr = matched[2];
 	} else {
-		portstr = str.substr(c + 1);
+		throw (new Error('no host specified'));
+	}
+
+	if (portstr !== undefined) {
+		mod_assert.equal(portstr[0], ':');
+		portstr = portstr.slice(1);
 		port = parseInt(portstr, 10);
-		host = str.substr(0, c);
 
-		if (isNaN(port))
-			throw (new Error('invalid port: "' + portstr + '"'));
-	}
+		if (Number(portstr) !== Math.floor(port)) {
```

Comment:
I'll add a `parseInteger()` function jsprim. I realized that this wouldn't cover the case where `portstr == "0x0"`, or, if using a newer node, `"0b0"` or `"0o0"`. (I also realized that the `Math.floor()` is unnecessary, since it's acting on the results of `parseInt()` and not `parseFloat()`.) This should cover all other cases where the string isn't a positive or negative integer in base 10, though.